File: LP3-DAA/pract_2_huffman.cpp
#include <iostream>
#include <bits/stdc++.h>
using namespace std;

class Node{
    public:
    bool is_leaf;
    int freq_sum;
    char c;
    Node* left;
    Node* right;
    Node(char c,int f){
        this->c = c;
        this->freq_sum = f;
        this->left = NULL;
        this->right = NULL;
        this->is_leaf = true;
    }

    Node(int f,Node* l=NULL,Node* r=NULL){
        this->is_leaf = false;
        this->freq_sum = f;
        this->left = l;
        this->right = r;
    }
};

void build_code_table(Node* root,string code,map<char,string>& code_table){
    if(root == NULL) return;
    if(root->is_leaf){
        code_table[root->c] = code;
        return;
    }

    build_code_table(root->left,code+'0',code_table);
    build_code_table(root->right,code+'1',code_table);
}


int main(){
    // input string
    string s;
    cout<<"Enter string To Encode:";
    cin>>s;

    // build frequency table
    map<char,int> freq;
    for(char c:s) freq[c]++;

    cout<<"Frequency Of Chars:"<<endl;
    for(auto p:freq){
        cout<<p.first<<" - "<<p.second<<endl;
    }

    // build huffman tree
    vector<Node*> nodes;
    for(auto p:freq){
        nodes.push_back(new Node(p.first,p.second));
    }
    sort(nodes.begin(),nodes.end(),[](Node* a,Node* b){return a->freq_sum < b->freq_sum;});

    while(nodes.size()>1){
        int sum = nodes[0]->freq_sum + nodes[1]->freq_sum;

        Node* parent = new Node(sum,nodes[0],nodes[1]);

        nodes.erase(nodes.begin(),nodes.begin()+2);

        nodes.push_back(parent);

        sort(nodes.begin(),nodes.end(),[](Node* a,Node* b){return a->freq_sum < b->freq_sum;});
    }

    Node* root = nodes[0];

    // build code table
    map<char,string> codes;
    string code_str = "";
    build_code_table(root,code_str,codes);

    // print code table
    cout<<"Code Table:"<<endl;
    for(auto p:codes){
        cout<<p.first<<" - "<<p.second<<endl;
    }

    // encode string
    string encoded_str = "";
    for(char c:s){
        encoded_str += codes[c];
    }

    cout<<"Encoded String:"<<endl;
    cout<<encoded_str<<endl;

    return 0;
}

File: DAA/pract_2_huffman.py
class node:
    def __init__(self):
        self.freq = 0
        self.symbol = ''
        self.left = ''
        self.right = ''
        self.is_leaf = False
        
    def create_leaf(self,symbol:str,freq:int)->None:
        self.is_leaf = True
        self.symbol = symbol
        self.freq = freq
        self.left = None
        self.right = None
        
    def create_node(self,freq,left,right):
        self.is_leaf = False
        self.freq = freq
        self.left = left
        self.right = right
        self.symbol = None
        
def insert_node(arr : list[node],node:node)->list[node]:
    i = len(arr)-1
    arr.append(node)
    while(i>=0 and arr[i].freq < node.freq):
        arr[i+1] = arr[i]
        i-=1
    arr[i+1] = node
    return arr

code_map = {}

def generate_huffman(node:node,code:str):
    global code_map
    if(node is None):
        return
    if(node.is_leaf):
        code_map[node.symbol] = code
        return
    generate_huffman(node.left,code+'0')
    generate_huffman(node.right,code+'1')
    

def main():
    s = input("Enter the string: ")
    m = {}
    for i in s:
        if i in m:
            m[i] += 1
        else:
            m[i] = 1
    # display frequency
    print("Frequency:")
    for i,j in m.items():
        print(i,":",j) 
    
    arr = []
    for i,j in m.items():
        n = node()
        n.create_leaf(i,j)
        arr = insert_node(arr,n)
        
    while(len(arr) > 1):
        n1 = arr.pop()
        n2 = arr.pop()
        n = node()
        n.create_node(n1.freq+n2.freq,n1,n2)
        arr = insert_node(arr,n)
        
    # generate codes
    generate_huffman(arr[0],'')
    
    # print codes
    print("Huffman Codes:")
    for i,j in code_map.items():
        print(i,":",j)
        
    # string encoding
    enc = ''
    for i in s:
        enc += code_map[i] + ' '
        
    print("Encoded String:")
    print(enc)

main()
